### PR TITLE
Design doc for the new Notification system

### PR DESCRIPTION
**Pull request by mithuna**
This design doc has details on the proposed Notification system and the
replacement of old system.

### CODE REVIEW

File: DESIGN.md [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+Previously we have been displaying notifications for each archive and hence the notifications feeds were also configured to fetch data based on the archive. Our design has never 
+considered providing users with an ability to take action right from the notification feed. In recent years, new features haven't been rolled out with accompanying notifications, 
+which has left the notifications feed limited in the types of notices it displays to users. 
+@Bryson new design has combined all feeds into one account notification feed, and also has actionable buttons corresponding to each feed. The limitations of the current 
+notification system calls for a redesign and re-architecture. 
+
+## Detailed Design
+
+We will create a new microsystem, preferably in TypeScript and a new database , the chosen one is PostgreSQL.  Microsystem will make it  easier to open source the new service in
```
Comment (jasonaowen) [inline DESIGN.md:26]: The term of art is [microservice](https://en.wikipedia.org/wiki/Microservices); let's search-and-replace all the occurrences of "microsystem". :slightly_smiling_face:

**Review by jasonaowen — CHANGES_REQUESTED** [review]
This is a great start! Thanks for writing this up, @mithuna.

I think we can polish this up a bit. I'd be happy to pair with you on this, and/or move to Google Docs to iterate; GitHub's code review tools aren't really well suited to technical writing.

File: DESIGN.md (hunk 1 above)
Comment (mithuna) [reply DESIGN.md:25]: > The term of art is [microservice](https://en.wikipedia.org/wiki/Microservices); let's search-and-replace all the occurrences of "microsystem". 🙂

Ah thank you.. My bad

**Comment by cecilia-donnelly** [comment]
(FYI I am not ignoring this, just waiting until the first round of changes are done to put in my review.)

**Comment by jasonaowen** [comment]
Thanks for making some of the suggested changes, @mithuna! Let me know if you disagree with or want to discuss any of the remaining changes, if you want to pair on writing.

File: DESIGN.md [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+## Objective
+
+To Create a new Notification system in TypeScript to provide users with better feeds and more features. 
+
+## Goals
+
+Our current notification feed has many limitations. It is specific to a selected archive and doesn’t include all the notifications generated in the system.
+Also our current notifications don’t have actionable buttons which can help users navigate easily in the app and save a lot of time.
+The proposed notification system will be able to eliminate all the limitations of the current notification system and also be able to provide users with better features like 
+actionable buttons.
```
[the remark this answers is not in the document]
Comment (cecilia-donnelly) [reply DESIGN.md:13]: Agree, especially since our current focus is on push notifications.

File: docs/DESIGN.md [hunk 3]
```diff
@@ -17,14 +17,14 @@ actionable buttons.
 Previously we have been displaying notifications for each archive and hence the notifications feeds were also configured to fetch data based on the archive. Our design has never
 considered providing users with an ability to take action right from the notification feed. In recent years, new features haven't been rolled out with accompanying notifications,
 which has left the notifications feed limited in the types of notices it displays to users.
-@Bryson new design has combined all feeds into one account notification feed, and also has actionable buttons corresponding to each feed. The limitations of the current
-notification system calls for a redesign and re-architecture.
+The new UI ![design](..\notification_ui_design.png) has combined all feeds into one account notification feed, and also has actionable buttons corresponding to each feed. The limitations of the current notification system calls for a redesign and re-architecture.
```
Comment (jasonaowen) [inline docs/DESIGN.md:20]: This link isn't working for me. It looks like `DESIGN.md` and `notification_ui_design.png` are both in `docs/` - which is good! - so you shouldn't need to change directories at all: `design`.

Once that is fixed, I think you'll find that dropping the image in-line like that is going to disrupt the flow. The image is 1600x1258, which is a bit large for text to just wrap around. Can you rephrase so that the text accommodates a large illustration?

File: README.md [hunk 4]
```diff
@@ -72,13 +72,18 @@ Returns a health check.
...
 - `toUserId`: Id of the user the notification is for.
 - `notificationType`: Type of the notification being saved.
 
 #### Output
+
 - `notificationId`: Id of the notification that got created.
 
+## Design Doc
+
+  link: [DESIGN.md](DESIGN.MD)
```
Comment (jasonaowen) [inline README.md:85]: This link is no longer correct:

```suggestion
  link: [DESIGN.md](docs/DESIGN.md)
```

File: docs/DESIGN.md [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+
+## Detailed Design
+
+We will create a new microservice, preferably in TypeScript and a new database , the chosen one is PostgreSQL.  Microservice will make it  easier to open source the new service in
+a usable way for other people. This will also offer a fresh codebase with correct process(es), preparing for a better-designed future. The endpoints will be directed through
+current API for authentication and authorization.  API will have 3 endpoints: one to fetch all notifications for the current user, one for the system to create a notification for
+a user, and one for a user to mark the notification as read; broadly, Create, Read, Update. Drawbacks of creating this as a microservice is that we need to have a new database
+limiting data integrity checking at db level.
+Users will also be provoded with options to set preferences for each set of notifications. This microservice will also send emails and push notifications. We are planning to use Firebase Cloud Messaging client for push notifications.
```
Comment (jasonaowen) [inline docs/DESIGN.md:29]: This is good additional detail, thank you!

File: docs/DESIGN.md [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+which has left the notifications feed limited in the types of notices it displays to users.
+The new UI ![design](..\notification_ui_design.png) has combined all feeds into one account notification feed, and also has actionable buttons corresponding to each feed. The limitations of the current notification system calls for a redesign and re-architecture.
+
+## Detailed Design
+
+We will create a new microservice, preferably in TypeScript and a new database , the chosen one is PostgreSQL.  Microservice will make it  easier to open source the new service in
+a usable way for other people. This will also offer a fresh codebase with correct process(es), preparing for a better-designed future. The endpoints will be directed through
+current API for authentication and authorization.  API will have 3 endpoints: one to fetch all notifications for the current user, one for the system to create a notification for
+a user, and one for a user to mark the notification as read; broadly, Create, Read, Update. Drawbacks of creating this as a microservice is that we need to have a new database
```
Comment (jasonaowen) [inline docs/DESIGN.md:27]: My [previous comment](https://github.com/PermanentOrg/notification-service/pull/4#discussion_r577186173) still applies; specifically, I would like for us to avoid stating details about endpoints. "API will have 3 endpoints" is not useful at a design level, and is not going to be correct for any length of time.

**Review by jasonaowen — CHANGES_REQUESTED** [review]
Thanks for working on this, @mithuna! It's looking better.

File: README.md [hunk 7]
```diff
@@ -72,13 +72,18 @@ Returns a health check.
...
 - `toUserId`: Id of the user the notification is for.
 - `notificationType`: Type of the notification being saved.
 
 #### Output
+
 - `notificationId`: Id of the notification that got created.
 
+## Design Doc
+
+  link: [DESIGN.md](docs/DESIGN.md)
```
Comment (mithuna) [inline README.md:85]: I am not able to navigate to this design doc. Is it because it is not merged? I can do this locally.

File: docs/DESIGN.md [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+Also our current notifications don’t have actionable buttons which can help users navigate easily in the app and save a lot of time.
+The proposed notification system will be able to eliminate all the limitations of the current notification system and also be able to provide users with better features like actionable buttons.
+
+## Overview
+
+Previously we have been displaying notifications for each archive and hence the notifications feeds were also configured to fetch data based on the archive. Our design has never considered providing users with an ability to take action right from the notification feed. In recent years, new features haven't been rolled out with accompanying notifications, which has left the notifications feed limited in the types of notices it displays to users.
+The new UI design has combined all feeds into one account notification feed, and also has actionable buttons corresponding to each feed. The limitations of the current notification system calls for a redesign and re-architecture. Below you can find the new UI ![design](notification_ui_design.png)
```
Comment (mithuna) [inline docs/DESIGN.md:17]: I have pushed this image file on a previous commit, but Git says it is empty.

File: README.md (hunk 7 above)
Comment (jasonaowen) [reply README.md:85]: Looking at the [version on your branch](https://github.com/PermanentOrg/notification-service/blob/PER-8245/README.md#design-doc), the link seems to work for me. Where are you trying to view it?

File: docs/DESIGN.md (hunk 8 above)
Comment (jasonaowen) [reply docs/DESIGN.md:17]: I saw that empty file in the PR diff too, but it's clearly not correct - maybe a bug in GitHub?

Looking at [DESIGN.md on your branch](https://github.com/PermanentOrg/notification-service/blob/PER-8245/docs/DESIGN.md), the image does show up!

Comment (mithuna) [reply docs/DESIGN.md:17]: Yes I can see it. May be I was looking at the wrong place.

File: README.md (hunk 7 above)
Comment (mithuna) [reply README.md:85]: Yes I can see it. May be I was looking at the wrong place.

**Review by jasonaowen — APPROVED** [review]
This looks good to me; thank you for working on it, @mithuna!

@cecilia-donnelly, did you have anything you wanted to add?

**Review by cecilia-donnelly — APPROVED** [review]
Thanks so much for writing this up, you two!
